File: 2023/01/javascript/main.js
const fs = require("fs");

const input = fs.readFileSync("../input.txt").toString().trim();

console.log(`Part 1: ${part1()}`);
console.log(`Part 2: ${part2()}`);

function part1() {
  return input
    .split("\n")
    .map((x) => [...x].map(Number).filter((c) => !isNaN(c)))
    .map((x) => {
      let f = x[0];
      let l = x[x.length - 1];
      return f * 10 + l;
    })
    .reduce((a, c) => a + c, 0);
}

function part2() {
  const nums = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ];

  return input
    .split("\n")
    .map((line) => {
      let first = nums.map((n, i) => [line.indexOf(n), i]).filter((x) => x[0] != -1);
      first.sort((a, b) => a[0] - b[0]);
      let last = nums.map((n, i) => [line.lastIndexOf(n), i]).filter((x) => x[0] != -1);
      last.sort((a, b) => b[0] - a[0]);
      let f = (first[0][1] % 9) + 1;
      let l = (last[0][1] % 9) + 1;
      return f * 10 + l;
    })
    .reduce((a, c) => a + c, 0);
}

File: 2023/01/rust/src/main.rs
fn main() {
  let input = std::fs::read_to_string("../input.txt").unwrap();
  let p1 = part1(&input);
  let p2 = part2(&input);
  println!("Part 1: {p1}");
  println!("Part 2: {p2}");
}

fn part1(input: &str) -> String {
  input
    .lines()
    .map(|line| {
      line
        .chars()
        .filter_map(|c| c.to_digit(10))
        .collect::<Vec<u32>>()
    })
    .map(|x| match x[..] {
      [] => 0,
      [o] => o * 10 + o,
      [f, .., l] => f * 10 + l,
    })
    .sum::<u32>()
    .to_string()
}

fn part2(input: &str) -> String {
  let nums = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "1", "2", "3", "4",
    "5", "6", "7", "8", "9",
  ]
  .iter()
  .enumerate();

  input
    .lines()
    .map(|line| {
      let first: usize = nums
        .clone()
        .filter_map(|(i, s)| line.match_indices(s).next().map(|(a, _)| (a, i)))
        .min_by_key(|(a, _)| *a)
        .unwrap()
        .1;

      let last = nums
        .clone()
        .filter_map(|(i, s)| line.rmatch_indices(s).next().map(|(a, _)| (a, i)))
        .max_by_key(|(a, _)| *a)
        .unwrap()
        .1;

      let f = first % 9 + 1;
      let l = last % 9 + 1;

      f * 10 + l
    })
    .sum::<usize>()
    .to_string()
}
